[MRG+1] Fixes return_X_y should be available on more dataset loaders/fetchers (#10734)

ccatalfo opened the pull request:
#### Reference Issues/PRs
Fixes #10734 by implementing return_X_y for kdcupp99, twenty_newsgroups, rcv1, and lfw datasets.

#### What does this implement/fix? Explain your changes.
This replicates the return_X_y parameter that was added to datasets/base.py.

#### Any other comments?
I did not add return_X_y to some of the other datasets as it seemed to make less sense for these.

ccatalfo committed:
[MRG] Update fetch_kddcup99 to support return_X_y (#10734)

ccatalfo committed:
[MRG] Update fetch_rcv1 to support return_X_y (#10734)

ccatalfo committed:
[MRG] Update fetch_20newsgroups_vectorized to support return_X_y (#10734)

ccatalfo committed:
[MRG] return_X_y to more datasets - pep8 (#10734)

ccatalfo committed:
[MRG] return_X_y to more datasets - versionadded fixed (#10734)

ccatalfo committed:
[MRG] return_X_y to more datasets - pep8 fixes (#10734)

ccatalfo committed:
[MRG] Update lfw to support return_X_y (#10734)

ccatalfo committed:
[MRG] Add return_X_y to twenty_newsgroups - fix versionadded (#10734)

jnothman commented on sklearn/datasets/tests/test_20news.py:
```diff
@@ -77,6 +78,13 @@ def test_20news_vectorized():
     assert_equal(bunch.data.dtype, np.float64)
 
+    # test return_X_y option
+    X_y_tuple = datasets.fetch_20newsgroups_vectorized(subset='test',return_X_y=True)
```
space after , please

jnothman commented on sklearn/datasets/tests/test_lfw.py:
```diff
@@ -139,6 +140,15 @@ def test_load_fake_lfw_people():
 
+    # test return_X_y option
+    X_y_tuple = fetch_lfw_people(data_home=SCIKIT_LEARN_DATA, resize=None,
+                                  slice_=None, color=True,
```
this indent should be consistent with the argument on the previous line

jnothman reviewed (commented):
This is still marked WIP. Is there something else you intend to do before considering this sufficient for review and merge?

ccatalfo committed:
[MRG] Add return_X_y to test_20news and test_lfw - fix indentation and spaces. (#10734)

ccatalfo commented:
@jnothman thanks for the heads up on those two issues and on marking as MRG.  I believe this is now ready for review and merge.

jnothman commented:
flake8 says you have left some lines longer than 79 characters.

jnothman reviewed (approved):
the tests are not run by our ci, I think. Have you run them?

It might be nice to have this as a common test for datasets. Otherwise LGTM

ccatalfo committed:
[MRG] Add return_X_y to other datasets  - fix some too-long lines. (#10734)

ccatalfo commented:
@jnothman I have run the tests and they are passing.   

I just committed fix for those too-long lines flagged by flake8.

A common test for datasets - meaning a ci test that runs the sklearn/datasets/tests ?

jnothman commented:
No, by a common test, I mean avoiding the repetition of code in the current tests, instead looping over fetches to confirm that they return the right format when return_X_y is used

ccatalfo committed:
[MRG] Add return_X_y to test_rcv1 - test shape rather than entire array to get around memory isssue. (#10734)

ccatalfo commented:
@jnothman "avoiding the repetition of code in the current tests" - Ah I see what you're saying.  Certainly could do that - does it make more sense to have every dataset's test in its own test_ file or test across datasets for functionality that crosses them, is the question I guess.

I made a small change to test_rcv1 where I think the test was running out of memory attempting to test the entire array returned.

The codecov/patch that is marked as failing at https://codecov.io/gh/scikit-learn/scikit-learn/compare/ccbf9975fcf1676f6ac4f311e388529d3a3c4d3f...7dcadcb12a74b4b871c1f4d976564992c25ce30a  - is that indicating the previous diff does not hit large enough test coverage percentage?

jnothman commented:
the codecov failure simply says that the new test code is not being run.
Which is to be expected.

a common test would live in datasets/tests/test_common.py.

ccatalfo commented:
@jnothman looking at the datasets/tests/test_common.py idea for return_X_y.

Perhaps I'm misinterpreting your idea, but I think there'd still wind up being duplicated code from moving the relevant pieces of the the test_<dataset>.py files into the tests/test_common.py:

While looping over the limited set of datasets which accept the return_X_y parameter (rcv1, lfw, 20_newsgroups, kddcup99 and various fetches from base.py)

1. test_common.py would have to catch exceptions from not having each dataset downloaded, as the normal test_lfw.py etc. test files do.

2. test_common.py would also have to call the particular dataset's Bunch-returning fetch function, each of which take different parameters, in order to compare the original Bunch with the X_y_tuple from returned from the return_X_y fetch.

In sum - while the actual **test** of the X_y_tuples are the same, the **fetching** involved differs by dataset.  Moving that to test_common.py would lead to code duplication of that part of the test logic.

While I agree that it would be nice to capture the repetitive parts of this return_X_y test logic, it feels to me like it would 

1. wind up having to duplicate nearly as much code as it replaces and
2. it would break up the coherence of having each dataset's tests only found in each test_<dataset>.py.

If I've misunderstood or mischaracterized your proposal, please let me know.   If you feel like test_common.py is the best way to go, I can certainly implement it that way.  Thanks for your thoughts.

Also thanks for the hand-holding as I get acclimated to the codebase and the contributing flow.

jnothman commented:
Okay. Could also just use a helper called check_return_X_y called by each
dataset's test.

ccatalfo commented:
@jnothman great, yes, I will do that - refactor the return_X_y test part called by each test into tests/test_common.py.

ccatalfo committed:
[MRG] Add return_X_y to more datasets - refactor tests of return_X_y parameter into common test function.

ccatalfo commented:
@jnothman I refactored as suggested into test_common.py - both the files I had updated (kddcup99, lfw etc) as well as test_base.py to use the new test function.

jnothman commented on sklearn/datasets/tests/test_common.py:
```diff
@@ -0,0 +1,10 @@
+
+
+def check_return_X_y(bunch, X_y_tuple):
+    assert_true(isinstance(X_y_tuple, tuple))
```
We are trying to phase out these assertion functions. Please use a bare assert

ccatalfo replied:
Also fixed in https://github.com/scikit-learn/scikit-learn/pull/10774/commits/5f73583c99c773fbc204b63afe2acce793fe58fa

jnothman commented on sklearn/datasets/tests/test_common.py:
```diff
@@ -0,0 +1,10 @@
+
+def check_return_X_y(bunch, X_y_tuple):
+    assert_true(isinstance(X_y_tuple, tuple))
+    assert_array_equal(X_y_tuple[0].shape, bunch.data.shape)
```
Shapes should not be arrays. Just use assert ==

ccatalfo replied:
Ok, switched in https://github.com/scikit-learn/scikit-learn/pull/10774/commits/5f73583c99c773fbc204b63afe2acce793fe58fa

jnothman commented on sklearn/datasets/tests/test_kddcup99.py:
```diff
@@ -38,6 +40,10 @@ def test_percent10():
     assert_equal(data.target.shape, (9571,))
 
+    X_y_tuple = fetch_kddcup99('smtp', return_X_y=True)
+    bunch = fetch_kddcup99('smtp')
```
Don't we already have a bunch above? I think it's unclear to the reader why we get this again

ccatalfo replied:
Thanks - you're right.   Fixed in https://github.com/scikit-learn/scikit-learn/pull/10774/commits/e68f3f99076b6de01354b5c9ef920075fa7a040f

jnothman commented on sklearn/datasets/tests/test_common.py:
```diff
@@ -0,0 +1,10 @@
+from sklearn.utils.testing import assert_true
+
+
+def check_return_X_y(bunch, X_y_tuple):
```
I had imagined this would take a partial function and pass in return_X_y=True. You still have left a lot of duplicated idiom in the test functions

ccatalfo replied:
By 

> a lot of duplicated idiom

do you mean the setting up of the bunches and fetch of the X_y_tuples to compare?

Were you thinking of something like this?

**Each of the dataset test files sets up the appropriate fetch partial**

* test_20news.py, test_lfw.py, etc.

```python
from functools import partial
# test return_X_y option
fetch_func = partial(datasets.fetch_20newsgroups_vectorized, subset='test')
check_return_X_y(bunch, fetch_func)
```

**During the check_return_X_y call, the partial is called, passing return_X_y=True** 

* test_common.py

```python
def check_return_X_y(bunch, fetch_func_partial):
    X_y_tuple = fetch_func_partial(return_X_y=True)
    assert(isinstance(X_y_tuple, tuple))
    assert(X_y_tuple[0].shape == bunch.data.shape)
    assert(X_y_tuple[1].shape == bunch.target.shape)
```

Wouldn't this run into problems if ever additional arguments are added to those fetch functions after what's currently the last parameter, the return_X_y one?

jnothman reviewed (commented):
Thanks!
You have a flake8 failure

ccatalfo committed:
[MRG] Add return_X_y to more datasets - remove unused import.

ccatalfo committed:
[MRG] Add return_X_y to more datasets - remove unneeded data fetch in test.

ccatalfo committed:
[MRG] Add return_X_y to more datasets - switch to straight asserts in test.

ccatalfo committed:
[MRG] Add return_X_y to more datasets - remove another unused import.

jnothman commented:
no, I don't see how it would run into problems if args are passed by name

ccatalfo commented:
Ok is that what you’re going for as far as the usage of the partial then? 

If so I’ll do that tonight.

jnothman commented:
I think so, only if it seems to make things neater

ccatalfo committed:
[MRG] Add return_X_y to more datasets - change check_return_X_y to use partial fetch function (#10734).

ccatalfo commented:
@jnothman Ok I've updated each to use the partial function version.  How does this look now?

jnothman reviewed (approved):
I'm happy with this :)

qinhanmin2014 commented on sklearn/datasets/tests/test_base.py:
```diff
@@ -177,11 +174,7 @@ def test_load_diabetes():
-    assert_true(isinstance(X_y_tuple, tuple))
-    assert_array_equal(X_y_tuple[0], bunch.data)
-    assert_array_equal(X_y_tuple[1], bunch.target)
+    check_return_X_y(res, partial(load_diabetes))
```
Why do we need partial here? (and other places in this file)

ccatalfo replied:
These tests_base.py test functions pass in the partial for the same reason the partial is passed in from the other test_* dataset files like test_20news.py - so that check_return_X_y can call that same dataset fetch function with the additional return_X_y=True parameter and perform the same standard X_y_tuple checks.

Doing it this way keeps the interface to check_return_X_y uniform among the dataset test_* files -- even if these particular fetch functions are relatively simple compared to the test files like test_20news.py.

qinhanmin2014 replied:
I don't want to argue on such a minor question (though I still prefer to remove partial if we don't need it). But I think you should try to make the whole file consistent. (e.g., you're not using partial in test_load_digits)

ccatalfo replied:
@qinhanmin2014 Aahh - good catch, I had missed that one.  Thanks.

ccatalfo replied:
I've pushed a fix for that.

qinhanmin2014 commented on sklearn/datasets/tests/test_20news.py:
```diff
@@ -77,6 +79,10 @@ def test_20news_vectorized():
     assert_equal(bunch.target.shape[0], 7532)
     assert_equal(bunch.data.dtype, np.float64)
 
+    # test return_X_y option
```
How about the training set?

ccatalfo replied:
I thought only one subset return_X_y check was needed since fetch_20newsgroups_vectorized is subsetting the fetched dataset before return_X_y is checked.

qinhanmin2014 reviewed (commented):
LGTM overall

qinhanmin2014 commented:
@ccatalfo Seems that some functions are not included here (e.g., ``fetch_california_housing``). Any specific reason?

ccatalfo commented:
@qinhanmin2014 As far as adding return_X_y to the last few datasets like california_housing it looked to me as if there were more than just the X and y to return.

For example `fetch_california_housing` returns 

`python
 return Bunch(data=data,
                 target=target,
                 feature_names=feature_names,
                 DESCR=MODULE_DOCS)`

so it didn't seem to make sense  to return just the `X=data` and `y=target` and leave out the `feature_names`.  But I could certainly add to that one if you think it's useful just with the X and y.

I'm not sure which other datasets, beyond that one, might benefit from return_X_y?  

Maybe covtype.py?  

So shall I add return_X_y tests to these two?
1. fetch_california_housing
2. covtype

jnothman commented:
yes, it's useful even without feature names.

ccatalfo committed:
[MRG] Add return_X_y to more datasets - add return_X_y to covtype dataset.

ccatalfo committed:
[MRG] Add return_X_y to more datasets - fix missing partial on load_digits test to check_return_X_y

ccatalfo commented:
I've added return_X_y to covtype.

ccatalfo committed:
[MRG] Add return_X_y to more datasets - add new test_california_housing datasets test file (#10374)

ccatalfo committed:
[MRG] Add return_X_y to more datasets - add return_X_y parameter to california_housing dataset (#10734).

ccatalfo commented:
Also added return_X_y to california_housing.

In doing that, I also added a new test_california_housing.py test file (I didn't see an existing test file for california_housing.py) and return_X_y.

qinhanmin2014 reviewed (approved):
LGTM. I can't fully understand the codecov failure but I think it's irrelevant and 'll trust jnothman's comment. Thanks @ccatalfo :)

ccatalfo commented:
Thanks for all the suggestions and comments!

jnothman commented:
thank you for your work and clear communication!
